Sources/SwiftFrameCore/Workers/ImageComposer.swift

```diff
@@ -0,0 +1,106 @@
+import AppKit
+import Foundation
+import CoreGraphics
+import CoreImage
+
+final class ImageComposer {
+
+    // MARK: - Properties
+
+    private let textRenderer = TextRenderer()
+    private let screenshotRenderer = ScreenshotRenderer()
+
+    let context: CGContext
+
+    // MARK: - Init
+
+    init(canvasSize: CGSize) throws {
+        self.context = try ImageComposer.createContext(size: canvasSize)
+    }
+
+    // MARK: - Preparation
+
+    private static func createContext(size: CGSize) throws -> CGContext {
+        guard let context = CGContext(
+            data: nil,
+            width: Int(size.width),
+            height: Int(size.height),
+            bitsPerComponent: 8,
+            bytesPerRow: 0,
+            space: CGColorSpaceCreateDeviceRGB(),
+            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue)
+        else {
+            throw NSError(description: "Failed to create graphics context")
+        }
+        return context
+    }
+
+    // MARK: - Composition
+
+    func addTemplateImage(_ image: NSBitmapImageRep) throws {
+        guard let templateImage = image.cgImage else {
+            throw NSError(description: "Could not render template image")
+        }
+
+        context.saveGState()
+        defer { context.restoreGState() }
+
+        context.draw(templateImage, in: image.ky_nativeRect)
+    }
+
+    // MARK: - Titles Rendering
+
+    func addStrings(_ associatedStrings: [AssociatedString], maxFontSizeByGroup: [String: CGFloat], font: NSFont, color: NSColor, maxFontSize: CGFloat) throws {
+        try associatedStrings.forEach {
+            if let sharedSize = $0.data.groupIdentifier.flatMap({ maxFontSizeByGroup[$0] }) {
+                // Can use fixed font size since common maximum has already been calculated
+                try add(
+                    title: $0.string,
+                    font: $0.data.fontOverride?.font() ?? font,
+                    color: $0.data.textColorOverride ?? color,
+                    fixedFontSize: sharedSize,
+                    textData: $0.data)
+            } else {
```

Comment:
Nit: It seems that the calls to `add` in `if` and `else` differ only by the font size arguments. Since the calls are lengthy, maybe extracting the font args to local variables and then just doing one call could end up being shorter.

Reply:
Just realized that it's two separate methods so this would only work if you create an enum for the font size specification (max or fixed) to make it into one method. Makes it even smaller a nit this way.